### PR TITLE
Flows shouldn't have to suspend if just doing a send

### PR DESCRIPTION
**Pull request by patrickkuo**
Jira - CORDA-45

### CODE REVIEW

File: node/src/main/kotlin/net/corda/node/services/statemachine/FlowStateMachineImpl.kt [hunk 1]
```diff
@@ -121,31 +121,38 @@ class FlowStateMachineImpl<R>(override val id: StateMachineRunId,
...
         return receiveInternal<SessionData>(session).checkPayloadIs(receiveType)
     }
 
     @Suspendable
     override fun send(otherParty: Party, payload: Any, sessionFlow: FlowLogic<*>) {
-        val (session, new) = getSession(otherParty, sessionFlow, payload)
-        if (!new) {
-            // Don't send the payload again if it was already piggy-backed on a session init
-            sendInternal(session, createSessionData(session, payload))
+        getSession(otherParty, sessionFlow)?.let {
```
Comment (shamsasari) [inline node/src/main/kotlin/net/corda/node/services/statemachine/FlowStateMachineImpl.kt]: Likewise

File: node/src/main/kotlin/net/corda/node/services/statemachine/FlowStateMachineImpl.kt [hunk 2]
```diff
@@ -121,31 +121,38 @@ class FlowStateMachineImpl<R>(override val id: StateMachineRunId,
...
+            sendInternal(it, createSessionData(it, payload))
+        } ?: startNewSession(otherParty, sessionFlow, payload)
+    }
+
+    /**
+     * This method will suspend the state machine and wait for incoming session init response from other party.
+     */
+    @Suspendable
+    private fun FlowSession.receiveSessionResponse() {
```
Comment (shamsasari) [inline node/src/main/kotlin/net/corda/node/services/statemachine/FlowStateMachineImpl.kt]: `waitForConfirmation`

File: node/src/main/kotlin/net/corda/node/services/statemachine/FlowStateMachineImpl.kt [hunk 3]
```diff
@@ -190,21 +197,17 @@ class FlowStateMachineImpl<R>(override val id: StateMachineRunId,
      * multiple public keys, but we **don't support multiple nodes advertising the same legal identity**.
      */
     @Suspendable
-    private fun startNewSession(otherParty: Party, sessionFlow: FlowLogic<*>, firstPayload: Any?): FlowSession {
+    private fun startNewSession(otherParty: Party, sessionFlow: FlowLogic<*>, firstPayload: Any?, waitForResponse: Boolean = false): FlowSession {
```
Comment (shamsasari) [inline node/src/main/kotlin/net/corda/node/services/statemachine/FlowStateMachineImpl.kt]: Can you rename `waitForResponse` to `waitForConfirmation` and not it give it a default value. Then use named parameters each time it's called (true/false param args are usually not descriptive)

File: node/src/main/kotlin/net/corda/node/services/statemachine/FlowStateMachineImpl.kt [hunk 4]
```diff
@@ -174,12 +181,12 @@ class FlowStateMachineImpl<R>(override val id: StateMachineRunId,
     }
 
     @Suspendable
-    private fun getSession(otherParty: Party, sessionFlow: FlowLogic<*>, firstPayload: Any?): Pair<FlowSession, Boolean> {
-        val session = openSessions[Pair(sessionFlow, otherParty)]
-        return if (session != null) {
-            Pair(session, false)
-        } else {
-            Pair(startNewSession(otherParty, sessionFlow, firstPayload), true)
+    private fun getSession(otherParty: Party, sessionFlow: FlowLogic<*>): FlowSession? {
```
Comment (shamsasari) [inline node/src/main/kotlin/net/corda/node/services/statemachine/FlowStateMachineImpl.kt]: `getConfirmedSession`

File: node/src/test/kotlin/net/corda/node/services/statemachine/StateMachineManagerTests.kt [hunk 5]
```diff
@@ -243,20 +243,21 @@ class StateMachineManagerTests {
 
         assertSessionTransfers(node2,
                 node1 sent sessionInit(SendFlow::class, payload) to node2,
-                node2 sent sessionConfirm to node1,
-                node1 sent sessionEnd to node2
+                node2 sent sessionConfirm to node1
+                //TODO: Node1 not sending back session end now because it didn't wait and receive the init response
```
Comment (shamsasari) [inline node/src/test/kotlin/net/corda/node/services/statemachine/StateMachineManagerTests.kt]: Session-end should be automatically sent once the flow finishes. I don't see how node1 not waiting for init-response should affect things.

Comment (patrickkuo) [inline node/src/test/kotlin/net/corda/node/services/statemachine/StateMachineManagerTests.kt]: node 1 won't have the session ID from other side if it doesn't receive the init-response; which is needed in the session end message...

Comment (shamsasari) [inline node/src/test/kotlin/net/corda/node/services/statemachine/StateMachineManagerTests.kt]: Of course. In that case the flow needs to wait for the session-confirm at the end if it hasn't received it.

Comment (patrickkuo) [inline node/src/test/kotlin/net/corda/node/services/statemachine/StateMachineManagerTests.kt]: the flow will now wait for session confirm at the end

File: node/src/main/kotlin/net/corda/node/services/statemachine/FlowStateMachineImpl.kt [hunk 6]
```diff
@@ -85,17 +85,25 @@ class FlowStateMachineImpl<R>(override val id: StateMachineRunId,
...
-
+        waitForAllSessionConfirmations()
         // This is to prevent actionOnEnd being called twice if it throws an exception
         actionOnEnd()
         _resultFuture?.set(result)
         return result
     }
 
+    @Suspendable
+    private fun waitForAllSessionConfirmations(){
```
Comment (shamsasari) [inline node/src/main/kotlin/net/corda/node/services/statemachine/FlowStateMachineImpl.kt]: Can you move this inside the `actionOnEnd` function as it's called both times before it

File: node/src/main/kotlin/net/corda/node/services/statemachine/FlowStateMachineImpl.kt [hunk 7]
```diff
@@ -121,34 +124,48 @@ class FlowStateMachineImpl<R>(override val id: StateMachineRunId,
...
         return receiveInternal<SessionData>(session).checkPayloadIs(receiveType)
     }
 
     @Suspendable
     override fun send(otherParty: Party, payload: Any, sessionFlow: FlowLogic<*>) {
-        val (session, new) = getSession(otherParty, sessionFlow, payload)
-        if (!new) {
+        val session = getConfirmedSession(otherParty, sessionFlow)
+        if (session == null) {
             // Don't send the payload again if it was already piggy-backed on a session init
```
Comment (shamsasari) [inline node/src/main/kotlin/net/corda/node/services/statemachine/FlowStateMachineImpl.kt]: Change this comment to say "piggy-back payload on session init" or similar

**Pull request merged**